mono/metadata/mono-security.c

```diff
@@ -337,15 +337,41 @@ ves_icall_System_Security_Principal_WindowsImpersonationContext_DuplicateToken (
 gboolean
 ves_icall_System_Security_Principal_WindowsImpersonationContext_SetCurrentToken (gpointer token)
 {
-	/* Posix version implemented in /mono/mono/io-layer/security.c */
+#ifdef HOST_WIN32
 	return (ImpersonateLoggedOnUser (token) != 0);
+#else
+	uid_t itoken = (uid_t) GPOINTER_TO_INT (token);
+#ifdef HAVE_SETRESUID
+	if (setresuid (-1, itoken, getuid ()) < 0)
+		return FALSE;
+#endif
+	return geteuid () == itoken;
+#endif
 }
 
 gboolean
 ves_icall_System_Security_Principal_WindowsImpersonationContext_RevertToSelf (void)
 {
-	/* Posix version implemented in /mono/mono/io-layer/security.c */
+#ifdef HOST_WIN32
 	return (RevertToSelf () != 0);
+#else
+#ifdef HAVE_GETRESUID
```

Comment:
Wouldn't it be better to have only one `HAVE_GETRESUID` which also contains the `getresuid` call and declare `suid` above the `#ifdef` ?